src/t/d19.clj

```diff
@@ -0,0 +1,219 @@
+(ns t.d19
+  (:gen-class)
+  (:require
+    [clojure.string :as cljstr]
+    [clojure.set :as cljset]
+    [clojure.core.match :refer [match]]))
+
+(def sample
+"r, wr, b, g, bwu, rb, gb, br
+
+brwrr
+bggr
+gbbr
+rrbgbr
+ubwu
+bwurrg
+brgr
+bbrgwb
+")
+(def sol 6)
+
+(defn parse-input
+  [input]
+  (let [[towels patterns] (cljstr/split input #"\n\n")
+        towels (cljstr/split towels #", ")
+        patterns (cljstr/split-lines patterns)]
+    [towels patterns]))
+
+
+(defn p-valid?
+  [p T]
+  (let [psize (count p)
+        t-exact (some (fn [t] (and
+                                (= psize (count t))
+                                (= t p))) T)
+        t-smaller (filter (fn [t] (and
+                                    (< (count t) psize)
+                                    (every? identity (map = t p))))
+                                    T)
+        ]
+;;    (prn [:p p :finish t-exact :t-smaller t-smaller])
+    (or
+      t-exact
+      (some
+        (fn[t] (p-valid? (drop (count t) p) T))
+        t-smaller)
+    ))
+  )
+
+(defn d19p1
+  [input]
+  (let [[T P] (parse-input input)
+        T (map vec T)
+        P (map vec P)
+        p (first P)
+        ]
+    (->> P
+         (keep (fn [p] (p-valid? p T)))
+         count)
+
+    ))
+
+
+
+(defn count-arr-M
+  [T]
+  (memoize
+    (defn count-arr
```

Comment:
It's bad form to do `defn` within a function call, because `defn` always defines a top-level binding. It's rarely what you want to do.

In this specific case you probably want `fn` instead, and, as discussed, you'd need to memoize the recursive call too.

Reply:
Also, `defn` returns the [Var](https://clojuredocs.org/clojure.core/var), not the function, so what's actually happening here is a bit subtle, though it ultimately does do what you want, I think:

```clojure
t.core=> (defn a [] (memoize (defn b [] (println "executed"))))
#'t.core/a
t.core=> ;; b exists already

t.core=> b
#object[clojure.lang.Var$Unbound 0x580a876d "Unbound: #'t.core/b"]
t.core=> (b)
Execution error (IllegalStateException) at t.core/eval9844 (form-init11052948005921262500.clj:1).
Attempting to call unbound fn: #'t.core/b

t.core=> (def c (a))
#'t.core/c
t.core=> (b)
executed
nil
t.core=> (b)
executed
nil
t.core=> (c)
executed
nil
t.core=> (c)
nil
t.core=> (c)
nil
t.core=>
```
So the `b` var exists as soon as `a` is defined, but it is unbound until `a` is called, and `b` is then the non-memoized version of `c`.

Reply:
Merci !
(y a tellement d'aspect de clj à apprendre ...)